torchtune/modules/peft/lora.py

```diff
@@ -130,6 +130,162 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
+            dtype=torch.int8,
+            granularity="per_token",
+            is_symmetric=False,
+        )
+        weight_qat_config = FakeQuantizeConfig(
+            dtype=torch.int4,
+            group_size=8,
+            is_symmetric=True,
+        )
+        qat_lora_linear = QATLoRALinear(
+            in_dim=512,
+            out_dim=1024,
+            rank=8,
+            alpha=16,
+            activation_qat_config=activation_qat_config,
+            weight_qat_config=weight_qat_config,
+        )
+        qat_lora_linear(torch.randn(512))
+    """
+
+    def __init__(
+        self,
+        in_dim: int,
+        out_dim: int,
+        rank: int,
+        alpha: float,
+        # fake quantize configs
+        # TODO: make the types Optional[FakeQuantizeConfig] once we
+        # support torchao 0.7+ by default
+        activation_qat_config: Optional["FakeQuantizeConfig"] = None,
+        weight_qat_config: Optional["FakeQuantizeConfig"] = None,
+    ):
+        super().__init__(
+            in_dim,
+            out_dim,
+            rank,
+            alpha,
+            dropout=0.0,
+            use_bias=False,
+            quantize_base=False,
+        )
+
+        try:
+            from torchao.quantization.qat.api import FakeQuantizeConfig
+            from torchao.quantization.qat.fake_quantizer import FakeQuantizer
+        except ImportError as err:
+            raise ValueError(
+                "QATLoRALinear is only compatible with torchao 0.7+"
+            ) from err
+
+        # initialize activation fake quantizer
+        if activation_qat_config is not None:
+            assert isinstance(activation_qat_config, FakeQuantizeConfig)
+            self.activation_fake_quantizer = FakeQuantizer(activation_qat_config)
+        else:
+            self.activation_fake_quantizer = None
+
+        # initialize weight fake quantizer
+        if weight_qat_config is not None:
+            assert isinstance(weight_qat_config, FakeQuantizeConfig)
+            group_size = weight_qat_config.group_size
+            if group_size is not None and in_dim % group_size != 0:
+                raise ValueError(
+                    "in_dim (%s) must be divisible by group_size (%s)"
+                    % (in_dim, group_size)
+                )
+            self.weight_fake_quantizer = FakeQuantizer(weight_qat_config)
+        else:
+            self.weight_fake_quantizer = None
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Args:
+            x (torch.Tensor): input tensor with shape ``(..., in_dim)``
+
+        Returns:
+            torch.Tensor: output tensor with shape ``(..., out_dim)``
+
+        """
+        if self.activation_fake_quantizer is not None:
+            _x = self.activation_fake_quantizer(x)
+        else:
+            _x = x
+        if self.weight_fake_quantizer is not None:
+            w = self.weight_fake_quantizer(self.weight)
+        else:
+            w = self.weight
```

Comment:
One thing we can consider to cut down on these if/elses in forward is just set activation_fake_quantizer and/or weight_fake_quantizer to nn.Identity() instead of None in the \_\_init\_\_